atomix/cluster/src/test/java/io/atomix/raft/roles/PassiveRoleTest.java

```diff
@@ -151,4 +152,44 @@ public void shouldNotFlushIfNoEntryIsAppended() {
     verify(log, never()).flush();
     assertThat(response.lastLogIndex()).isEqualTo(0);
   }
+
+  @Test
+  public void shouldStoreLastFlushedIndex() {
```

Comment:
Instead of this test, you can add a more high level test using `RaftRule` to create a cluster with 1 and 3 nodes. You can read the lastFlushedIndex directly from the metastore and verify it has been updated.

Reply:
I don't see why. This is shorter in scope and checks that the index is stored at (more or less) the correct time, that's not as easy to do with a high level test

Reply:
A high level test is preferable because it tests the behavior, and is less dependent on the internal implementation. So it would be better if we test - after restart it reads the correct flushedIndex, or after restart it is able to detect corruption. For example, in this case a higher level test that tests flushed index is updated after a successful append is testing the same behavior. I don't see any advantage of this test that mocks a lot of things.
I would prefer to have only the test for detecting corruption (once it is implemented). It does not really matter if or when the flushedIndex is updated, right?

Reply:
> A high level test is preferable because it tests the behavior

This tests the behavior. The behavior isn't just "it persists the index", the correct behavior is persisting the index before the follower responds to the append request. That is what this tests is testing for and it wouldn't be tested in a high level test (or at least not as simply)

> or after restart it is able to detect corruption.

I agree that this is important but this PR isn't dealing with corruption, only with persisting the index. I added that test in the commit that actually deals with partial writes/corruption and it'll be in the next PR

> It does not really matter if or when the flushedIndex is updated, right?

It matters that we persist the index before acknowledging the append request to the leader. Otherwise, the follower could fail after acknowledging and before persisting it, in which case, we couldn't distinguish between the last entry being corrupted or partially written

Reply:
> It matters that we persist the index before acknowledging the append request to the leader. Otherwise, the follower could fail after acknowledging and before persisting it, in which case, we couldn't distinguish between the last entry being corrupted or partially written

That's a valid point. But I do feel that the test is too implementation dependent. But let's keep it since I don't see a better alternative.